### PR TITLE
[SYCL][CUDA] Support host-device memcpy2D 

### PR DESCRIPTION
**Pull request by abagusetty**
Addresses to support host-device memcpy2D copies

### CODE REVIEW

File: sycl/unittests/kernel-and-program/DeviceInfo.cpp [hunk 1]
```diff
@@ -168,6 +169,9 @@ TEST_F(DeviceInfoTest, GetDeviceFreeMemory) {
 
   EXPECT_EQ(sizeof(FreeMemory), sizeof(uint64_t))
       << "Expect free_memory to be of uint64_t size";
+
+  EXPECT_GE(TotalMemory, FreeMemory)
+      << "Expect total_memory >= free_memory";
```
Comment (steffenlarsen) [inline sycl/unittests/kernel-and-program/DeviceInfo.cpp:174]: How is this related to your changes? I am not against it, but it seems like something that should have its own patch.

File: sycl/plugins/cuda/pi_cuda.cpp [hunk 2]
```diff
@@ -163,6 +163,8 @@ pi_result check_error(CUresult result, const char *function, int line,
 
 /// \cond NODOXY
 #define PI_CHECK_ERROR(result) check_error(result, __func__, __LINE__, __FILE__)
+// Use (void) to silence unused warnings.
+#define assertm(exp, msg) assert(((void)msg, exp))
```
Comment (steffenlarsen) [inline sycl/plugins/cuda/pi_cuda.cpp:167]: I am not sure I understand the need for this. The case where it is used do not seem have any potentially unused variables, and if the string is a problem, we normally just use something like `assert(predicate && "YOUR STRING HERE");`.

File: sycl/unittests/kernel-and-program/DeviceInfo.cpp (hunk 1 above)
Comment (abagusetty) [reply sycl/unittests/kernel-and-program/DeviceInfo.cpp:174]: This test is not related. I will move this into a separate patch.

File: sycl/plugins/cuda/pi_cuda.cpp (hunk 2 above)
Comment (abagusetty) [reply sycl/plugins/cuda/pi_cuda.cpp:167]: Thanks, I was exactly trying for predicate + a string . Will fix that.

File: sycl/plugins/cuda/pi_cuda.cpp [hunk 3]
```diff
@@ -5425,39 +5424,59 @@ pi_result cuda_piextUSMEnqueueMemcpy2D(pi_queue queue, pi_bool blocking,
...
+      cpyDesc.srcDevice = 0;
+      cpyDesc.srcHost = src_ptr;
+      // todo: resets the above "non-sticky" error
+    } else if (ret == CUDA_SUCCESS) {
+      cpyDesc.srcMemoryType = src_type;
+      cpyDesc.srcDevice = (src_type == CU_MEMORYTYPE_DEVICE)
+                              ? reinterpret_cast<CUdeviceptr>(src_ptr)
+                              : 0;
+      cpyDesc.srcHost = (src_type == CU_MEMORYTYPE_HOST) ? src_ptr : nullptr;
+    }
```
Comment (steffenlarsen) [inline sycl/plugins/cuda/pi_cuda.cpp:5452]: Since this behavior is becoming more complex, would it make sense to split it into a helper function so we can avoid repeating the code? I can see two ways of doing it:

1. Implement it as a function returning a triple, i.e. something like:
```c++
template<typename PtrT>
static std::tuple<CUmemorytype, CUdeviceptr, PtrT> getUSMHostOrDevicePtr(PtrT usm_ptr) { ... }
...
std::tie(cpyDesc.srcMemoryType, cpyDesc.srcDevice, cpyDesc.srcHost) = getUSMHostOrDevicePtr(src_ptr);
...
std::tie(cpyDesc.dstMemoryType, cpyDesc.dstDevice, cpyDesc.dstHost) = getUSMHostOrDevicePtr(dst_ptr);
```
2. Do the same, but make it `void` and pass the members to set as pointers/references:
```c++
template<typename PtrT>
static void getUSMHostOrDevicePtr(PtrT usm_ptr, CUmemorytype *out_mem_type, CUdeviceptr *out_dev_ptr, PtrT *out_host_ptr) { ... }
...
getUSMHostOrDevicePtr(src_ptr, &cpyDesc.srcMemoryType, &cpyDesc.srcDevice, &cpyDesc.srcHost);
...
getUSMHostOrDevicePtr(dst_ptr, &cpyDesc.dstMemoryType, &cpyDesc.dstDevice, &cpyDesc.dstHost);
```
Note I use pointers above because I like to think it makes the intention clearer here, but I do not have a strong preference for one or the other.

Comment (abagusetty) [reply sycl/plugins/cuda/pi_cuda.cpp:5452]: Sure, I agree.

File: sycl/plugins/cuda/pi_cuda.cpp [hunk 4]
```diff
@@ -5425,39 +5424,59 @@ pi_result cuda_piextUSMEnqueueMemcpy2D(pi_queue queue, pi_bool blocking,
...
+    // do not throw if cuPointerGetAttribute returns CUDA_ERROR_INVALID_VALUE
+    // checks with PI_CHECK_ERROR are not suggested
+    CUresult ret = cuPointerGetAttribute(
+        &src_type, CU_POINTER_ATTRIBUTE_MEMORY_TYPE, (CUdeviceptr)src_ptr);
+    assert(
+        (src_type != CU_MEMORYTYPE_ARRAY && src_type != CU_MEMORYTYPE_UNIFIED) &&
+        "ARRAY, UNIFIED types are not supported!");
+
+    // pointer not known to the CUDA subsystem (possibly a system allocated ptr)
+    if (ret == CUDA_ERROR_INVALID_VALUE) {
```
Comment (jchlanda) [inline sycl/plugins/cuda/pi_cuda.cpp:5441]: Is it safe to check only for two possible error codes from `cuPointerGetAttribute`?

**Comment by abagusetty** [comment]
/verify with https://github.com/intel/llvm-test-suite/pull/1597

**Review by jchlanda — APPROVED** [review]
LGTM

**Review by steffenlarsen — APPROVED** [review]
Looks good!
